### PR TITLE
feat(msgraph): support advanced querying capabilities

### PR DESCRIPTION
**Pull request by pjungermann**
Support advanced querying capabilities
using the new config option

```
queryMode?: 'basic' | 'advanced'
```

which will set all required headers and
query parameters.

Closes: #9795
Signed-off-by: Patrick Jungermann <Patrick.Jungermann@gmail.com>

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/catalog-backend-module-msgraph/src/microsoftGraph/read.ts
Comment (pjungermann) [inline plugins/catalog-backend-module-msgraph/src/microsoftGraph/read.ts:535]: this was missing before and could have been extracted into a separate change.

**Comment by ehrnst** [comment]
thanks this is a great addition :)

**Comment by pjungermann** [comment]
@ehrnst feel free to review the changes, too

**Review by ehrnst — COMMENTED** [review]
This looks OK with me. Not sure how the review-routine works with Backstage.

**Comment by pjungermann** [comment]
@ehrnst I believe if it looks ok to you, you can leave an approval. Maintainers will have a look, too. But that helps them if there was already another review from the community.

File: .changeset/beige-lamps-cry.md
Comment (freben) [inline .changeset/beige-lamps-cry.md:2]: should be minor since it's a breaking change to a public interface

File: plugins/catalog-backend-module-msgraph/README.md
Comment (freben) [inline plugins/catalog-backend-module-msgraph/README.md:40]: So this is still useful to be able to pass in, even though we can detect the mode for `search` as advanced, right?

**Review by freben — COMMENTED** [review]
alright, let's just update the changeset and then good to go!

File: plugins/catalog-backend-module-msgraph/README.md
Comment (pjungermann) [reply plugins/catalog-backend-module-msgraph/README.md:40]: yes, as certain values for "filter" will require advanced mode, too, and parsing and detecting this is way to complex and hard to maintain. Hence, we leave it to the user.

You could say we do an auto-upgrade to advanced if you use certain features like "search".

We could fail if search is used without using advanced, but that would be a breaking change.

File: .changeset/beige-lamps-cry.md
Comment (pjungermann) [reply .changeset/beige-lamps-cry.md:2]: I guess you mean the change to the client.getUsers and client.getGroups functions, right @freben? Besides that queryMode is optional with backwards compatible default value applied at a low level.

If it is about these two functions, maybe a change of the position would reduce it again to "patch"? 🤔

Comment (freben) [reply .changeset/beige-lamps-cry.md:2]: I think it would, yes

File: plugins/catalog-backend-module-msgraph/README.md
Comment (freben) [reply plugins/catalog-backend-module-msgraph/README.md:40]: 👌

**Comment by freben** [comment]
@pjungermann just a small ping here :)

File: .changeset/beige-lamps-cry.md
Comment (pjungermann) [reply .changeset/beige-lamps-cry.md:2]: @freben I moved the new optional parameter to the end

**Review by freben — APPROVED** [review]
Aight!

**Pull request merged**
